pkg/server/local_observer.go

```diff
@@ -92,27 +114,202 @@ func (s *LocalObserverServer) GetStopped() chan struct{} {
+	// it will read all values from the given interval:
+	// [ lastWrite, s.ring.write [
+	lastWrite := ring.LastWriteParallel() + 1
+	readIdx := lastWrite - req.Number
+
+	switch {
+	case req.Follow:
+		ch = ring.ReadFrom(ctx.Done(), readIdx)
+	case req.Number != 0:
+		var ctx1 context.Context
+		ctx1, stop = context.WithCancel(ctx)
+		ch = ring.ReadFrom(ctx1.Done(), readIdx)
+	default:
+		beginning, err := types.TimestampFromProto(req.GetSince())
+		if err != nil {
+			return nil, nil, err
+		}
+		end, err := getUntil(req, types.TimestampNow())
+		if err != nil {
+			return nil, nil, err
+		}
+		timestampCh := make(chan *v1.Event, 1000)
+		ch = timestampCh
+
+		var ctx1 context.Context
+		ctx1, stop = context.WithCancel(ctx)
+
+		go func() {
+			defer close(timestampCh)
+			for lastWrite := ring.LastWriteParallel(); ; lastWrite-- {
+				e, ok := ring.Read(lastWrite)
+				// if the buffer was not full yet we can get nil payloads
+				if e == nil || e.Event == nil || !ok {
+					return
+				}
+				ts, err := types.TimestampFromProto(e.GetFlow().GetTime())
+				if err != nil {
+					return
+				}
+				if beginning.Before(ts) && end.After(ts) {
+					select {
+					case <-ctx1.Done():
+						return
+					case timestampCh <- e:
+					}
+				}
+			}
+		}()
+	}
+	return ch, stop, nil
 }
 
-// HandleMonitorSocket is a noop for local server since it doesn't connect to the monitor socket.
-func (s *LocalObserverServer) HandleMonitorSocket(nodeName string) error {
-	return nil
+// getFlow returns the flows either depending on the requests performed.
+func getFlows(
+	ctx context.Context,
+	log *zap.Logger,
+	ring *container.Ring,
+	req *observer.GetFlowsRequest,
+) (chan *pb.Flow, error) {
+	start := time.Now()
+	i := uint64(0)
+	defer func() {
+		size := ring.Cap()
+		took := time.Now().Sub(start)
+		log.Debug(
+			"GetFlows finished",
+			zap.Uint64("number_of_flows", i),
+			zap.Uint64("buffer_size", size),
+			zap.String("whitelist", logFilters(req.Whitelist)),
+			zap.String("blacklist", logFilters(req.Blacklist)),
+			zap.Duration("took", took),
+		)
+	}()
+
+	whitelist, err := filters.BuildFilterList(req.Whitelist)
+	if err != nil {
+		return nil, err
+	}
+	blacklist, err := filters.BuildFilterList(req.Blacklist)
+	if err != nil {
+		return nil, err
+	}
+
+	log.Debug("filters", zap.String("req", fmt.Sprintf("%+v", req)))
+	log.Debug("whitelist", zap.String("whitelist", fmt.Sprintf("%+v", whitelist)))
+	log.Debug("blacklist", zap.String("blacklist", fmt.Sprintf("%+v", blacklist)))
+
+	ch, stop, err := getBufferCh(ctx, ring, req)
+	if err != nil {
+		return nil, err
+	}
+	reply := make(chan *pb.Flow, 1)
+	go func() {
+		defer close(reply)
+		defer stop()
+
+		for e := range ch {
+			if req.Number != 0 && !req.Follow {
+				i++
+				if i >= req.Number {
+					// stop the channel buffer because we have reached
+					// the number of requested flows.
+					stop()
+					if i > req.Number {
+						// We will 'continue' since 'ch' might have flows and we
```

Comment:
I don't see a `continue` here, I see a return 🤔

Reply:
🤔 it came from https://github.com/cilium/hubble/pull/69/commits/fceda102a528ee3c3cae18b0da5dfc5921a4f7f3#diff-69945783af04950df9270c268c24c6a1L439 we can fix it in a follow-up commit. i prefer to keep the original code for a refactor commit.